### DISCUSSION TITLE
Application startup problems when running tests via the @QuarkusTest annotation

### DISCUSSION THREAD

**Discussion by aleks-iv4** [quarkusio/quarkus#28501] [Q&A] [answered]
Describe the bug
Reservation: I'm not sure if this's a bug or a work limitation of the annotation @QuarkusTest. This is our first big project on this framework, so please don't judge strictly.

My team and I write a project on Kotlin, using Quarkus as the main framework, and Gradle as the main build system. We have an application that we are currently covering with tests. The problem arises when we write REST-http tests to our EndPoints. The tests work out correctly, but only when we run them, with the application running. But if we turn off the application and put the @QuakusTest annotation, which, according to the documentation, should run the application itself before running the tests, an exception occurs:

"Connection refused: no further information
java.net.ConnectException: Connection refused: no further information"

Although the application itself is running.
We think that the problem may arise from working with GraphQL Schema.

We understand that requests should be sent to http://localhost:8081/mapins but it doesn't work. We get such response: "GraphQL Schema not generated. Make sure you have a GraphQL Endpoint. Go to https://quarkus.io/guides/microprofile-graphql to learn how". When we exclude the path from requests for autoconfiguration we get the same result as in the case when we send requests to http://localhost:8081/ - responses are just empty. We have already tried using quarkus.http.test-port but result are the same... Maybe the problem is that our application is multimodule and needed to be started from concrete module and @QuarkusTest uses some kind of default command?

This is a link (https://github.com/Bazalii/Interactive-map) to our small test project on github, on which we tested the work of a real project.

Output of java -version
Java 17, Kotlin 1.6.21

Quarkus version or git rev
2.11.1.Final

Build tool (ie. output of mvnw --version or gradlew --version)
Gradle 7.4.2

Additional information
We searched Internet for information about this problem, but the only thing we found was a bit similar question on StackOverflow, where the solution was - to connect "io.quarkus:quarkus-resteasy-reactive-jackson". We tried this option, but it didn't help(

Our tests:
![image](https://user-images.githubusercontent.com/70962483/195093958-02089690-ad37-49dc-998b-5c301ad4ae57.png)

Application properties:
![image](https://user-images.githubusercontent.com/70962483/195094068-c52aa99e-c6dd-4b7d-bace-f34ee126ae12.png)

Error when trying to run tests without an enabled application on a standard port:
![image](https://user-images.githubusercontent.com/70962483/195094553-46429b47-4f44-47da-a343-a90f0308f99c.png)

Error when trying to run tests without the application enabled on port 8080:
![image](https://user-images.githubusercontent.com/70962483/195094709-094c6c0f-25e9-4f2b-89fc-53286cab9c8a.png)
![image](https://user-images.githubusercontent.com/70962483/195094870-551a5c2a-3b18-4cc0-8c2a-9111ea1ebdcd.png)

Tests when the web application is running from the module:
![image](https://user-images.githubusercontent.com/70962483/195095250-32b9570b-2145-4fda-b86f-08788fad9e3f.png)

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Hi,

Isn't this a duplicate of https://github.com/quarkusio/quarkus/issues/28448?

**Reply by Bazalii** [reply]
Hi!

Yes, but right now it's placed in correct topic because you said that it's not the issue with quarkus but our problem. Unfortunately, we didn't get wanted answer there except proposal to read articles and check projects which we've already seen and processed. So, we decided to put it here because it seems to me that answers here are more meaningful and maybe someone else would help us with our problem. 

Sorry for tagging again, didn't thought that area is bound to tagging same people here.

**Reply by geoand** [reply]
No need to apologize, just asking to make sure we are on the same page :)

**Comment by geoand** [comment, marked as answer]
To reiterate, I'll just say that `@QuarkusTest` **always** starts the Quarkus application and also configured RestAssured with the base URL of the application.

@Sgitario is there anything more you would like to add here?

**Reply by Bazalii** [reply]
@geoand could you take us deeper inside how `@QuarkusTest` actually works? Does it use something like ./gradlew quarkusDev with port 8081 exposing? We have multimodule quarkus application that should be started from correct module(./gradlew web:quarkusDev) and all the example projects have src directory with all files in it. Can we change the command that `@QuarkusTest` uses to start the application?

**Reply by geoand** [reply]
It starts Quarkus in test mode, which similar to dev-mode but not exactly the same (for one thing it starts the application on port `8081` by default). It shuts down Quarkus when the tests are done.

**Reply by Bazalii** [reply]
@geoand Is it possible to configure(or manually change) the command with which @QuarkusTest starts the application?

**Reply by geoand** [reply]
Generally, it is not. What do you want to configure?

**Reply by Bazalii** [reply]
I want to set the module from which the application should be started during tests. We have multimodule apllication with core, web, data and tests modules. When I manually start the application I use command: ./gradlew web:quarkusDev so that all the dependencies between modules are resolved successfully. If I use just ./gradlew quarkusDev it starts the apllication using core module. I think the situation with QuarkusTest annotation is the same. It tries to start the application but it does not know how to do that properly.

**Reply by geoand** [reply]
@aloubyansky @glefloch I thought the Quarkus Gradle plugin did the right thing in the situation described above. Do you have any comments?

**Reply by aloubyansky** [reply]
It starts the app from the module being tested. I'll try https://github.com/Bazalii/Interactive-map

**Reply by aloubyansky** [reply]
I see that the `core` subproject fails to build, which is a dependency of the tests, perhaps that's the reason the app fails to start for the tests and needs to be fixed first.
````
[aloubyansky@localhost Interactive-map]$ ./gradlew clean build
The Agroal dependency is present but no JDBC datasources have been defined.
Hibernate ORM is disabled because no JPA entities were found
> Task :core:quarkusBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: Found 2 deployment problems: 
  [1] Unsatisfied dependency for type map.repositories.IMapRepository and qualifiers [@Default]
        - java member: map.services.implementations.MapService():_repository
        - declared on CLASS bean [types=[map.services.implementations.MapService, map.services.IMapService, java.lang.Object], qualifiers=[@Default, @Any], target=map.services.implementations.MapService]
  [2] Unsatisfied dependency for type users.repositories.IUserRepository and qualifiers [@Default]
        - java member: users.services.implementations.UserService():_userRepository
        - declared on CLASS bean [types=[users.services.implementations.UserService, java.lang.Object, users.services.IUserService], qualifiers=[@Default, @Any], target=users.services.implementations.UserService]
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1217)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:275)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:134)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:494)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:977)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
        Suppressed: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type map.repositories.IMapRepository and qualifiers [@Default]
        - java member: map.services.implementations.MapService():_repository
        - declared on CLASS bean [types=[map.services.implementations.MapService, map.services.IMapService, java.lang.Object], qualifiers=[@Default, @Any], target=map.services.implementations.MapService]
                at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:411)
                at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:532)
                at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:263)
                ... 13 more
        Suppressed: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type users.repositories.IUserRepository and qualifiers [@Default]
        - java member: users.services.implementations.UserService():_userRepository
        - declared on CLASS bean [types=[users.services.implementations.UserService, java.lang.Object, users.services.IUserService], qualifiers=[@Default, @Any], target=users.services.implementations.UserService]
                at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:411)
                at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:532)
                at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:263)
                ... 13 more
````

**Reply by aloubyansky** [reply]
No, it does start in tests
````
MapControllerHttpTests STANDARD_ERROR
    Nov 21, 2022 8:48:21 AM io.quarkus.bootstrap.runner.Timing printStartupTime
    INFO: tests 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.11.1.Final) started in 2.979s. Listening on: http://localhost:8081
    Nov 21, 2022 8:48:21 AM io.quarkus.bootstrap.runner.Timing printStartupTime
    INFO: Profile test activated. 
    Nov 21, 2022 8:48:21 AM io.quarkus.bootstrap.runner.Timing printStartupTime
    INFO: Installed features: [agroal, cdi, flyway, hibernate-orm, hibernate-orm-panache-kotlin, jdbc-h2, jdbc-postgresql, kotlin, narayana-jta, resteasy-reactive, resteasy-reactive-jsonb, security, security-jpa, smallrye-context-propagation, smallrye-graphql, vertx]

MapControllerHttpTests > getChairById_SuccessPath_ChairFoundAndReturnedCorrectStatusCode() FAILED
````

**Reply by aloubyansky** [reply]
Here you go @Bazalii 
````
diff --git a/gradlew b/gradlew
old mode 100644
new mode 100755
diff --git a/tests/src/test/kotlin/map/MapControllerHttpTests.kt b/tests/src/test/kotlin/map/MapControllerHttpTests.kt
index 3f3cb6b..7f88fa5 100644
--- a/tests/src/test/kotlin/map/MapControllerHttpTests.kt
+++ b/tests/src/test/kotlin/map/MapControllerHttpTests.kt
@@ -13,7 +13,7 @@ import java.util.*
 
 @QuarkusTest
 class MapControllerHttpTests {
-    private val _baseURL = "http://localhost:8080/mapins"
+    private val _baseURL = "http://localhost:8081/mapins"
 
     private fun generateStringFromResource(path: String): String {
         return String(Files.readAllBytes(Paths.get(path)))
diff --git a/tests/src/test/kotlin/users/UserControllerHttpTests.kt b/tests/src/test/kotlin/users/UserControllerHttpTests.kt
index da40897..a84733a 100644
--- a/tests/src/test/kotlin/users/UserControllerHttpTests.kt
+++ b/tests/src/test/kotlin/users/UserControllerHttpTests.kt
@@ -10,7 +10,7 @@ import java.util.*
 
 @QuarkusTest
 class UserControllerHttpTests {
-    private val _baseURL = "http://localhost:8080/mapins"
+    private val _baseURL = "http://localhost:8081/mapins"
````

**Reply by Bazalii** [reply]
> t starts the app from the module being tested

The problem here is that core module is isolated and it doesn't know about data project. All the dependencies are resolved correctly only when the application is started or built from web project(./gradlew web:quarkusBuild or ./gradlew web:quarkusDev)

**Reply by Bazalii** [reply]
@aloubyansky could you please clarify how did you start tests? Or please describe the whole subsequence of things that you have done to start tests? Beacuse when I just change port and start tests using ./gradlew tests:quarkusTests all the tests still fail

**Reply by aloubyansky** [reply]
./gradlew clean test

They do fail but for a different reason, something about uuid not matching.

**Reply by aloubyansky** [reply]
> > t starts the app from the module being tested
> 
> The problem here is that core module is isolated and it doesn't know about data project. All the dependencies are resolved correctly only when the application is started or built from web project(./gradlew web:quarkusBuild or ./gradlew web:quarkusDev)

Then it shouldn't be built as an application.

**Reply by Bazalii** [reply]
> ./gradlew clean test
> 
> They do fail but for a different reason, something about uuid not matching.

The problem is still the same. Application for tests is not active and as a result we get as a response 404 and not 200 as expected:

MapControllerHttpTests > getAllWalls_SuccessPath_AllWallsFoundAndCorrectStatusCodeIsReturned() FAILED
    java.lang.AssertionError at MapControllerHttpTests.kt:30

![image](https://user-images.githubusercontent.com/58232044/203028173-ad4294e8-09ac-4b44-aa31-75dafd2573fd.png)

**Reply by Bazalii** [reply]
> ./gradlew clean test
> 
> They do fail but for a different reason, something about uuid not matching.

The problem with UUID is that we get such response and not the correct one: 
![image](https://user-images.githubusercontent.com/58232044/203030061-46ac71e6-136e-42fd-bfac-5c5d7896c052.png)

**Reply by aloubyansky** [reply]
> > ./gradlew clean test
> > They do fail but for a different reason, something about uuid not matching.
> 
> The problem is still the same. Application for tests is not active and as a result we get as a response 404 and not 200 as expected:
> 
> MapControllerHttpTests > getAllWalls_SuccessPath_AllWallsFoundAndCorrectStatusCodeIsReturned() FAILED java.lang.AssertionError at MapControllerHttpTests.kt:30

It means the URL does not respond with the expected code/content but it's not because the app wasn't bootstrapped. You can see in the logs the app did start w/o errors.

The next fix would be to add `META-INF/beans.xml` to the `web` subproject (the same way it is added to the `core`).

**Reply by Bazalii** [reply]
@aloubyansky adding META-INF/beans.xml to the web worked for me! Thank you very much!

**Reply by aloubyansky** [reply]
@Bazalii here is what you need to know about the `@QuarkusTest` to get this cleared, `@QuarkusTest` will use the module (or subproject in Gradle's terms) the test belongs to as the *main* application's module. In your case, the `web` subproject is *not* the main module of the application you are testing, it's a dependency of the test application. As a consequence, to make the beans and other stuff available in your test application, you need to add `META-INF/beans.xml` to the `web` subproject.

**Reply by Bazalii** [reply]
> @Bazalii here is what you need to know about the `@QuarkusTest` to get this cleared, `@QuarkusTest` will use the module (or subproject in Gradle's terms) the test belongs to as the _main_ application's module. In your case, the `web` subproject is _not_ the main module of the application you are testing, it's a dependency of the test application. As a consequence, to make the beans and other stuff available in your test application, you need to add `META-INF/beans.xml` to the `web` subproject.

Okay, thank you for explanations and for your attentive help with this problem! It cannot be described in words for how long we have tried to solve the problem and the solution was just so simple..
